chore(hybrid): Move extra context to standalone class

krystofwoldrich opened the pull request:
## :scroll: Description

<!--- Describe your changes in detail -->

This enables us to reuse the same extra context logic in hybrid SDKs.

## :bulb: Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

- https://github.com/getsentry/sentry-react-native/issues/2884

## :green_heart: How did you test it?

cocoa sample app and rn sample app

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps
#skip-changelog

brustolin commented on Sources/Sentry/SentryExtraContextProvider.h:
```diff
@@ -0,0 +1,18 @@
+
++ (instancetype)sharedInstance;
+
+- (_Nonnull instancetype)init;
```
You dont need this. Init without arguments is the default one.

```suggestion
```

brustolin commented on Sources/Sentry/SentryExtraContextProvider.m:
```diff
@@ -0,0 +1,75 @@
+
+- (_Nonnull instancetype)init
+{
+    self.crashWrapper = [SentryCrashWrapper sharedInstance];
+    self.deviceWrapper = [[SentryUIDeviceWrapper alloc] init];
+    self.processInfoWrapper = [[SentryNSProcessInfoWrapper alloc] init];
+    return self;
```
```suggestion
if (self = [super init]) {
    self.crashWrapper = [SentryCrashWrapper sharedInstance];
    self.deviceWrapper = [[SentryUIDeviceWrapper alloc] init];
    self.processInfoWrapper = [[SentryNSProcessInfoWrapper alloc] init];
}
    return self;
```

brustolin commented on Sources/Sentry/SentryExtraContextProvider.m:
```diff
@@ -0,0 +1,77 @@
+
+- (NSDictionary *)getExtraContext
+{
+    NSMutableDictionary *extraContext = [[NSMutableDictionary alloc] init];
+
+    [extraContext setValue:[self getExtraDeviceContext] forKey:@"device"];
+    [extraContext setValue:[self getExtraAppContext] forKey:@"app"];
+
+    return extraContext;
```
l: 
```suggestion
    return @{ @"device":[self getExtraDeviceContext], @"app":[self getExtraAppContext] };
```

brustolin commented on Sources/Sentry/SentryExtraContextProvider.m:
```diff
@@ -0,0 +1,77 @@
+
+- (NSDictionary *)getExtraAppContext
+{
+    NSMutableDictionary *extraAppContext = [[NSMutableDictionary alloc] init];
+    extraAppContext[SentryDeviceContextAppMemoryKey] = @(self.crashWrapper.appMemorySize);
+    return extraAppContext;
```
l:
```suggestion
    return @{ SentryDeviceContextAppMemoryKey: @(self.crashWrapper.appMemorySize)};
```

brustolin commented on Sources/Sentry/SentryExtraContextProvider.m:
```diff
@@ -0,0 +1,77 @@
+    return instance;
+}
+
+- (_Nonnull instancetype)init
```
l: 
```suggestion
- (instancetype)init
```

brustolin commented on Sources/Sentry/SentryExtraContextProvider.m:
```diff
@@ -0,0 +1,77 @@
+- (_Nonnull instancetype)init
+{
+    if (self = [super init]) {
+        self.crashWrapper = [SentryCrashWrapper sharedInstance];
+        self.deviceWrapper = [[SentryUIDeviceWrapper alloc] init];
+        self.processInfoWrapper = [[SentryNSProcessInfoWrapper alloc] init];
```
m: I think this 3 properties should be parameters in the init, easier to make tests.

brustolin commented on Sources/Sentry/SentryClient.m:
```diff
@@ -144,9 +142,9 @@ - (instancetype)initWithOptions:(SentryOptions *)options
                           locale:[NSLocale autoupdatingCurrentLocale]
-                        timezone:[NSCalendar autoupdatingCurrentCalendar].timeZone];
+                        timezone:[NSCalendar autoupdatingCurrentCalendar].timeZone
+            extraContextProvider:extraContextProvider];
```
m: instead of requesting an `extraContextProvider` here, I think you should rollback the init functions, and initialize extraContextProvider inside the init, and provide it with the needed wrappers as mentioned in this comment, then you wont need to fix the tests.

brustolin reviewed (commented):
First full look.
Some minor suggestions to make code cleaner. 

And I think, instead of providing extraContextProvider to SentryClient, you should make `SentryContextProvider` init to ask for `crashWrapper`, `deviceWrapper` and `ProcessInfoWrapper`, then you initialize a SentryContextProvider inside SentryClient.
This way you fix CI

krystofwoldrich commented:
@brustolin Thank you for the suggestions and help with the test. I've added tests for the provider and fixed the client tests. I hope the CI will be happy this time.

brustolin commented on Sources/Sentry/SentryExtraContextProvider.h:
```diff
@@ -0,0 +1,23 @@
+
++ (instancetype)sharedInstance;
+
+- (_Nonnull instancetype)initWithCrashWrapper:(SentryCrashWrapper *)crashWrapper
```
l: nonnull is already assumed because of line 6.

```suggestion
- (instancetype)initWithCrashWrapper:(SentryCrashWrapper *)crashWrapper
```

brustolin reviewed (approved):
LGTM
